include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -756,23 +767,15 @@ __parallel_transform_reduce_then_scan(oneapi::dpl::__internal::__device_backend_
                                       _ScanInputTransform __scan_input_transform, _WriteOp __write_op, _InitType __init,
                                       _Inclusive, _IsUniquePattern)
 {
-    using _ValueType = typename _InitType::__value_type;
     using _CustomName = oneapi::dpl::__internal::__policy_kernel_name<_ExecutionPolicy>;
-    // Note that __sub_group_size and __max_inputs_per_item are not included in kernel names. __sub_group_size
-    // is always constant (32) and __max_inputs_per_item is directly tied to the input type so these are not
-    // necessary to obtain a unique kernel name. However, if these compile time variables are adjusted in the
-    // future, then we need to be careful here to ensure unique kernel naming.
-    using _ReduceKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<
-        __reduce_then_scan_reduce_kernel, _CustomName, _ExecutionPolicy, _InRng, _OutRng, _GenReduceInput, _ReduceOp,
-        _InitType, _Inclusive, _IsUniquePattern>;
-    using _ScanKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<
-        __reduce_then_scan_scan_kernel, _CustomName, _ExecutionPolicy, _InRng, _OutRng, _GenScanInput, _ReduceOp,
-        _ScanInputTransform, _WriteOp, _InitType, _Inclusive, _IsUniquePattern>;
-    static auto __kernels = __internal::__kernel_compiler<_ReduceKernel, _ScanKernel>::__compile(__exec);
-    const sycl::kernel& __reduce_kernel = __kernels[0];
-    const sycl::kernel& __scan_kernel = __kernels[1];
-
-    constexpr std::uint8_t __sub_group_size = 32;
+    using _ReduceKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+        __reduce_then_scan_reduce_kernel<_CustomName>>;
+    using _ScanKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<
+        __reduce_then_scan_scan_kernel<_CustomName>>;
+    using _ValueType = typename _InitType::__value_type;
+
+    // To workaround hardware bugs on older integrated graphics, we need to use a sub-group size of 16 with -O0
+    // compilation.
```

Comment:
If we do change `__reduce_then_scan_sg_sz` to a constexpr function, we could call it here and assign it to `constexpr __sub_group_size`, and leave `__sub_group_size` as the variable in the rest.

Reply:
Done